connection.go

```diff
@@ -103,11 +107,21 @@ Loop:
 				}
 			}()
 
+			go func() {
+				for _ = range setTimeoutChan {
+				}
+			}()
+
 			i.writeLock.Lock()
```

Comment:
Is there a reason not to grab both locks and clear both channels in the same section here?  Seems like a race waiting to happen to me, but I am very inexperienced in this codebase and no specific risks jump out.  Just a gut feeling.